### PR TITLE
feat: add pixelization filter of an image

### PR DESCRIPTION
**Pull request by EscapedGibbon**

### CODE REVIEW

**Comment by stropitek** [comment]
The method should throw if the conditions for the cell size are not met: an integer >= 2

File: src/filters/pixelate.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+export function pixelate(image: Image, options: PixelateOptions): Image {
+  const { cellSize } = options;
+  const newImage = getOutputImage(image, options);
+
+  for (let channel = 0; channel < image.channels; channel++) {
+    for (let column = 0; column < image.width; column += cellSize) {
+      for (let row = 0; row < image.height; row += cellSize) {
+        const currentCellWidth = Math.min(cellSize, image.width - column);
+        const currentCellHeight = Math.min(cellSize, image.height - row);
+        //first case: image gets pixelated without any small parts remaining
```
Comment (stropitek) [inline src/filters/pixelate.ts:46]:
```suggestion
```

**Comment by stropitek** [comment]
@EscapedGibbon can you run `git config --global pull.rebase true`?

It will rebase by default and avoid those kind of merge commits when you pull:

CleanShot 2023-05-16 at 09 06 56

**Review by lpatiny — APPROVED** [review]
Looks good to me. Just fix the error message.

File: src/filters/pixelate.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ *Function to pixelate an image
+ *
+ * @param image - image to be pixelated
+ * @param options - PixelateOptions
+ * @returns pixelated Image
+ */
+export function pixelate(image: Image, options: PixelateOptions): Image {
+  const { cellSize } = options;
+  if (cellSize < 2) {
+    throw new Error('invalid option value. cellSize should be bigger than 2');
```
[the remark this answers is not in the document]
Comment (targos) [reply src/filters/pixelate.ts:40]: We use "must be" when it's a hard requirement.

[the remark this answers is not in the document]
Comment (targos) [reply src/filters/pixelate.ts:40]: We should also throw a `RangeError` here,

**Comment by stropitek** [comment]
The integer check is missing

File: src/filters/pixelate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (targos) [reply src/filters/pixelate.ts:40]: oh and finally, we don't say "invalid option value". Here's how it should be:

`throw new RangeError('cellSize must be greater than 1');`

and for the integer check (to be placed before):

`throw new TypeError('cellSize must be an integer');`

File: src/filters/__tests__/pixelate.test.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    let result = pixelate(img, { cellSize: 3 });
+
+    expect(result).toMatchImageData([
+      [2, 3, 4, 2],
+      [2, 3, 4, 2],
+      [2, 3, 4, 2],
+      [2, 9, 4, 0],
+      [2, 9, 4, 0],
+    ]);
+  });
```
Comment (stropitek) [inline src/filters/__tests__/pixelate.test.ts:74]: duplicate test (exact same one as above)

File: src/filters/__tests__/pixelate.test.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    let result = pixelate(img, { cellSize: 3 });
+
+    expect(result).toMatchImageData([
+      [2, 3, 4, 2],
+      [2, 3, 4, 2],
+      [2, 3, 4, 2],
+      [2, 9, 4, 0],
+      [2, 9, 4, 0],
+    ]);
```
Comment (stropitek) [inline src/filters/__tests__/pixelate.test.ts:92]: ditto

**Comment by EscapedGibbon** [comment]
Done

**Pull request merged**